### ISSUE TITLE
[feature request] Support dynamic metadata spec in State Management

### ISSUE THREAD

**Issue by olitomlinson** [dapr/dapr#4993]
We have data security requirements in our multi-tenant system whereby we need to :

1. Control how data is logically segregated by `tenant_id` when persisted.
2. Prevent developers from accidentally (or intentionally) omitting important query constraints that would restrict the result set to a given `tenant_id`.

Dapr State Management building blocks currently don't provide a way to achieve this dynamically at runtime for an unbounded amount of tenants, which makes it incredibly challenging, if not impossible, for us to adopt State Management in most of our use-cases.

However, if the State Management building block could become _tenant-aware_ and provide directives to the underlying components on _how_ to store/retrieve the data that may enable us to utilise State Management.

---

Given a State Management component named `widgetStore` 
When I reference the `widgetStore` component in my user code, I would like to be able to pass a known `tenant_id` property, like so

```c#
var tenantId = "fb057741-8c7e-4968-8ef2-197c396c7f2a";
var DAPR_STORE_NAME = $"widgetStore_tenant:{tenant_id}";
await client.SaveStateAsync(DAPR_STORE_NAME, "widget_1", myWidget.ToString());
```

Given the component name of `widgetStore_tenant:fb057741-8c7e-4968-8ef2-197c396c7f2a` , dapr would detect additional data in the name, and resolve the correct component `widgetStore` and tokenize the `tenant` property and value.

The `tenant` property would propagate to the underlying state store implementation and the become available as a variable to be dynamically interpolated into the metadata spec _at runtime._

### Example : MS-SQL component

Let's assume I'm using the MS-SQL implementation, I wanted the tenant data to be split by `Table`.
In the metadata spec for the component, I would reference the `tenant` property like so 

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: widgetStore
spec:
  type: state.sqlserver
  version: v1
  metadata:
  - name: tableName
    value: {tenant}_widgetStore
```

This would result in a SQL Table called `dbo.fb057741-8c7e-4968-8ef2-197c396c7f2a_widgetStore`

Alternatively, I could chose to segregate the tenant data by `schema`

```yaml
  metadata:
  - name: schema
    value: {tenant}_widgetStore
```

This would result in a SQL table called `fb057741-8c7e-4968-8ef2-197c396c7f2a_widgetStore.default`

### Example : Azure Blob Storage

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: widgetStore
spec:
  type: state.azure.blobstorage
  version: v1
  metadata:
  - name: containerName
    value: "{tenant}_widgetStore"
```

which would result in a path to the `widget_1` blob similar to `fb057741-8c7e-4968-8ef2-197c396c7f2a_widgetStore/widget_1`

---

1. I've provided Azure examples above because thats what seems to resonate the most with the community, but in reality I would utilise this with PostgresQL and S3 State Management components, but I don't actually think it matters for the sake of this proposal.
2. I'm aware that embedding dynamic properties into the component name programmatically might not be desirable from a maintainers point of view. I'm not trying to recommend this as the actual solution, I'm just trying to articulate the overall concept. It would probably be better to pass properties via HTTP header rather than embedding in the component name.
3. A design goal of this is that tenants are a _runtime_ and _unbounded_ concepts, so an alternative solution where many thousands of component yamls are created (i.e. one per tenant) would not be feasible.

## In what area(s)?

> /area runtime

**Comment by yaron2** [comment]
Thanks for this proposal @olitomlinson, I've discussed this with fellow maintainer @artursouza and we agree this is a very useful feature for Dapr, and can be extended to other building blocks in the future.

In terms of design, we propose the following (using HTTP as an example for now but this is applicable to gRPC also):

We'd like to introduce the concept of `tenant` as a standard and consistent sub-feature within building blocks.
This is because we want to ensure the following:

1. Consistency in usage across different APIs
2. Enable this for a larger number of components that don't necessarily surface tables/containers through metadata

To do this, we propose that a tenant is provided to the runtime as metadata:

```
POST v1.0/state/redis?metadata.tenant=fb057741-8c7e-4968-8ef2-197c396c7f2a
```

The Dapr runtime would then pass the metadata transparently to the component (as today). If the component supports the tenant feature, it will prefix the `bucket/table/container/other` name with the provided tenant. By default, the most granular part of storage will be used for tenancy. In the case of the SQL component you've shown above, this would be `table`.

If a component is modified in such a way that it supports the tenant feature on both a table, and, for example, a schema - the user will be able to change the latter in the following way:

```
POST v1.0/state/redis?metadata.schema.tenant=fb057741-8c7e-4968-8ef2-197c396c7f2a
```

**Comment by olitomlinson** [comment]
@yaron2 @artursouza Thanks for accepting the proposal! 

What you’ve said all makes sense to me.

I’ve another consideration to add to the mix, which is, can the specification of the tenant be mandated by the component yaml?

This is important because we absolutely wouldn’t want developers to accidentally miss this property. So if the Operator can enforce that the tenant is required (via a setting in the component yaml for example) then I would expect any operations against that state store to return a 400 bad request if tenant wasn’t specified by the developer when making the request.

**Comment by olitomlinson** [comment]
> If a component is modified in such a way that it supports the tenant feature on both a table, and, for example, a schema - the user will be able to change the latter in the following way:

> POST v1.0/state/redis?metadata.schema.tenant=fb057741-8c7e-4968-8ef2-197c396c7f2a

also, I somewhat disagree with this approach for my use-case.

I don’t want the developer to be able to change the strategy by providing `metadata.schema` or `metadata.table` - for me this is very much an Operations concern which a dev shouldn’t be able to modify at whim.

Not saying that this approach would work for other use-cases, but just not mine :)

**Comment by yaron2** [comment]
>  if the Operator can enforce that the tenant is required (via a setting in the component yaml for example) then I would expect any operations against that state store to return a 400 bad request if tenant wasn’t specified by the developer when making the request.

That makes sense to me. What about this type of component enforcement:

```
metadata:
  - name: tableName
     value: widgetStore
  - name: tenant
     value: tableName
```

This would make runtime operations more performant as we wouldn't need complex string manipulations. cc @artursouza

**Comment by olitomlinson** [comment]
Yep, that feels stricter and therefore better.

@andihysondotmatics thoughts?

**Comment by artursouza** [comment]
I agree to have  `?metadata.tenant` and not have anything that exposes the component spec to the user. If you need to enforce tenant in a component, it can be part of the component metadata (not request metadata):
```yaml
metadata:
  - name: tableName
    value: widgetStore
  - name: tenant
    value: tableName
  - name: multiTenant
    value: true
```

This way, the component code will check for the `tenant` value in the request metadata.

**Comment by yaron2** [comment]
> I agree to have `?metadata.tenant` and not have anything that exposes the component spec to the user. If you need to enforce tenant in a component, it can be part of the component metadata (not request metadata). Something like `multiTenant` = `true/false`:
> 
> ```yaml
> metadata:
>   - name: tableName
>     value: widgetStore
>   - name: tenant
>     value: tableName
>   - name: multiTenant
>     value: true
> ```
> 
> This way, the component code will check for the `tenant` value in the request metadata.

That seems fine. @artursouza might `multiTenant` be redundant and we could rely on the existence of a `tenant` property alone?

**Comment by artursouza** [comment]
I would let it be something defined per component.

> > I agree to have `?metadata.tenant` and not have anything that exposes the component spec to the user. If you need to enforce tenant in a component, it can be part of the component metadata (not request metadata). Something like `multiTenant` = `true/false`:
> > ```yaml
> > metadata:
> >   - name: tableName
> >     value: widgetStore
> >   - name: tenant
> >     value: tableName
> >   - name: multiTenant
> >     value: true
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This way, the component code will check for the `tenant` value in the request metadata.
> 
> That seems fine. @artursouza might `multiTenant` be redundant and we could rely on the existence of a `tenant` property alone?

I would let it be decided per component.

**Comment by andihysondotmatics** [comment]
This all looks great everyone - and thanks for considering the proposal.  

I think I'd agree with @yaron2 that the `multiTenant` aspect could be implied by the presence of the `tenant` property in most cases - however I could see where there could be certain kinds of component that have just one tenancy model.  In that case @artursouza's suggestion of a simple boolean works better than the key/value on `tenant` that can only have one possible value.

**Comment by olitomlinson** [comment]
Ideally, we would like this to be supported on the PostgresQL component as a first pass. 

Possibly S3 too, but less certain about this. Also, there isn't an S3 State Management component yet - I suspect it may make sense to introduce an S3 state management component once the [streaming capability in State Management](https://github.com/dapr/dapr/issues/4934) is available

**Comment by saber-wang** [comment]
We also have similar requirements, and hope to further expand to other building blocks, such as pubsub

**Comment by olitomlinson** [comment]
Hey, so I went ahead and built a pluggable component as a proof-of-concept to explore the idea of data segregatin strategies in the State Management building block, using postgres as the underlying store.

https://github.com/olitomlinson/tenant-aware-dapr-pluggable-state-store

I'm sharing what I learned to help direct/shape this feature. _This is not an attempt to share feedback on pluggable state stores, I've already done that via Discord._

**tldr** : on the whole, the proposal feels viable, but I've only assessed this from a postgres perspective.

---

**1. metadata schema**

As per our chats earlier in this proposal, I went ahead and implemented based on the following metadata schema

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: prod-mystore
spec:
  type: state.my-component
  version: v1
  metadata:
  - name: connectionString
    value: "<REDACTED>"
  - name: tenant
    value: schema | table
```

The addition of the new `tenant` metadata property to control if tenant-scoped Schemas/Tables were included worked well. No apparent design/usability issues with it.

**2. When the `tenant` property is specified in component metadata, all state store calls must include the accompanying `metadata.tenantId` property on each operation, otherwise error-out / 500**

I've validated this design decision and it works well.

**3. It's easier to create database objects eagerly rather than reactively.**

In my first build, I would perform a row-level Upsert/Delete operation, **assuming** that the target tenant-scoped Schema / Tenant-scoped table, would exist. If it didn't exist, the db would naturally throw an error, I would catch this in the app code and then go about creating the missing objects (which could be the tenant-scoped schema, or the tenant-scoped table, or both). The code got messy real quickly. I went down this route initially because I was biasing towards performance - I didn't want to check if the necessary database objects existed before I performed any row-level Upsert/Delete operations.

I rewrote the code to take a more predictable and linear code path. Every State Management request now followed this strategy :

`Create Schema` > `Create Table` > `Row Upsert/Delete operation`

- Check to see if the tenant-scoped schema exists (if schema tenancy was enabled)
- if not exists, create the tenant-scoped schema (or the default schema)
- Then, check to see if the tenant-scoped table exists (if table tenancy was enabled)
- if not exists, create the tenant-scoped table (or the default table)
- Finally, when we can be sure all pre-requisite db objects exists, perform the row-level Upsert/Delete operation.

Therefore...

**2. Any given implementation _may_ have to make a decision to trade-off performance vs simpler code-paths.**

In the long term it may be worth having an metadata component options which can hint to the implementation wether to use a more aggressive/performant mode, vs a simpler mode which is safer by virtue of eagerly checking the necessary pre-requisites exist before every call.

From what I've experienced, simplicity won over raw performance here.

**3. Things get hairy with `BulkSet` because dapr provides no expectations to surround the entire request in a transaction or not. It is left to the component implementation to decide, however this is not documented anywhere which has caused a portability conflict of State Management implementations**

- Redis **does not wrap** a transaction around the entire `BulkSet` request.
- Postgresql **does wrap** a transaction around the entire `BulkSet` request.
- Not investigated any others, so who knows?!

This gets more interesting when each operation in the `BulkSet` array targets _different_ `metadata.tenantId` - see example below.

```json

[{
   "key": "1",
   "value": { 
	"name" : "Dave Mustaine" 
   },
   "metadata": { 
	"tenantId" : "1" 
   }
},{
   "key": "2",
   "value": { 
	"name" : "Kirk Hammett" 
   },
   "metadata": { 
	"tenantId" : "2"
   }
}]

```

This forced me to think about the various error conditions that could occur. For example, Let's assume that the first operation in the `BulkSet` succeeded, but what if the second operation got into a partial or full error state - where would that leave the entire `BulkSet` operation? Should I rollback the first operation too, or should I leave it as-is.

I decided to wrap each `BulkSet` operation in the array, in a transaction. This would allow me to ensure that we wouldn't be left with orphaned tenant-scoped Schemas or orphaned tenant-scoped Tables in the event that the row-level Upsert/Delete operation failed.

Due to postgres not supporting nested transactions, I couldn't wrap a transaction around the _entire_ `BulkSet` request.

This led me to provide feedback to the dapr maintainers that the `BulkSet` API should set an expectation to State Management component authors that the `BulkSet` request _shouldn't_ be wrapped in a transaction. This would then complement the `Transactional API` nicely which _does_ mandate a transaction surrounding the entire request. 

Unfortunately, if my feedback was actioned today into the built-in postgres component, this would introduce a breaking change. So this may need to be addressed with an opt-in flag/config option to correct it's behaviour to be compliant to a as-to-be-yet documented expectation that entire `BulkSet` request shouldn't be wrapped in a transaction.

**4. Be aware of obsolete Tenant-scoped tables.**

In any system that runs for a sufficient length of time, it's reasonable to assume that some Tables/Schemas will be become dormant as tenants are deleted from a system.

Therefore, should Dapr State Management API extend to include a `Delete Tenant` operation? Which would in turn DROP the tenant-scoped Table and/or tenant-scoped Schema?

Personally this `Delete Tenant` capability would be valuable to our use-case, but not essential from day one.

**5. Similar to ` Delete Tenant` from # 4, offer an accompanying `Create Tenant` endpoint.**

In an evented system (a new customer signs up and a tenant is created), it may be possible to create the Schemas/Tables in the State Store ahead of time before they are needed. This would pair well with more aggressive form of the strategy as mentioned in # 3 where Upserts/Deletes don't eagerly check if the schema/table exists as an assumption is made that they already do exist.

**Comment by olitomlinson** [comment]
@yaron2

Assuming we're happy with _how_ we pass a tenant Id via meta data

Could you see this working for a Redis-backed KV State Store?

The `tenant` is prefixed/injected into the `key`

for example : 

`namespace-1.my-app||key1`

becomes

`namespace-1.tenant-123.my-app||key1`

---

Or does redis offer a better more suitable virtual container/collection that is more akin to a database table or schema?

**Comment by StephenStrickland** [comment]
Looking forward to this effort, just offering my teams use case. CosmosDB for document storage, tenant data nested within the document, code enforced isolation.

**Comment by olitomlinson** [comment]
@StephenStrickland Thanks for sharing! 

When you say _code enforced isolation_, does this mean that all of your queries append an extra constraint that specifically narrows the scope of the query to just documents which match that particular tenant property? 

Do you share the same concerns as mine which is that a developer may accidentally forget to apply that tenant constraint when querying and accidentally retrieve records that belong to a different tenant?

Also, just curious, would you ever be open to the idea of isolating each tenants data into their own collection? I know there used to be some challenges around cost here (After a certain number of collections, you were charged an addition 100RUs per additional collection, or something like that....) but my knowledge is a couple of years out of date here so things may have changed.

**Comment by StephenStrickland** [comment]
>When you say code enforced isolation, does this mean that all of your queries append an extra constraint that specifically narrows the scope of the query to just documents which match that particular tenant property?

>Do you share the same concerns as mine which is that a developer may accidentally forget to apply that tenant constraint when querying and accidentally retrieve records that belong to a different tenant?

Exactly this, we have reusable code that provides the current request's tenant information, and then that get's passed into our data access layer. There's 100% a concern there that something could get missed, but we have tests to validate that.

>Also, just curious, would you ever be open to the idea of isolating each tenants data into their own collection?

Definitely thought about this, but it was easier to optimize cost by using the same container (tradeoff being noisy neighbor problem). Testing is also easier bc you have a single container to talk to without the need to hunt down a specific tenants container.

We've also been looking into the hierarchical partition keys. Seem like a killer feature, can't wait for it to be GA.

**Comment by olitomlinson** [comment]
Bump

**Comment by arnoldpistorius** [comment]
We also have a multi-tenant application that needs a separate container per tenant on Azure blob. This feature would make our lives much easier.

**Comment by artursouza** [comment]
I tagged it as a good first issue. We can accept this from the community if done prior to the next release but removing it from the planned items.

**Comment by olitomlinson** [comment]
User on discord expressed a requirement for multi-tenant data segregation for various state store components

User requested : 

> Azure Blob and Cosmos would be my first preference. S3 and dynamo second.

My quick response is that this could be achieved in a couple of ways :

**Azure blobs** 
table per tenant (physical)
key prefix (logical)

**cosmos**
container per tenant (physical)
partition key prefix (logical)

**S3**
path per tenant (logical)
bucket per tenant (physical)

**Dynamo**
key prefix per tenant (logical)
not sure if dynamo has a more aggressive/physical tier?



There could be a light weight solution to all of those if logical is the minimum requirement here, as this wouldn't mean creating runtime objects in each store (tables/buckets/containers etc)

However, it puts us into a grey area if that logical segregation is enough - some customers will say it is, some will say it isn't

https://discord.com/channels/778680217417809931/1245741620138344458
